### ISSUE TITLE
Undefined variable when using generic type parameters (PEP695)

### ISSUE THREAD

**Issue by Manitary** [pylint-dev/pylint#9193]
### Bug description

The behaviour described in #9110 is still observable with pylint 3.0.2:

```python
def f[T](a: T) -> T: ...
```

works fine but `T` is marked as undefined variable

### Command used

```shell
pylint .\mwe.py
```

### Pylint output

```python
************* Module mwe
mwe.py:1:0: C0114: Missing module docstring (missing-module-docstring)
mwe.py:1:0: C0116: Missing function or method docstring (missing-function-docstring)
mwe.py:1:12: E0602: Undefined variable 'T' (undefined-variable)
mwe.py:1:18: E0602: Undefined variable 'T' (undefined-variable)
mwe.py:1:9: W0613: Unused argument 'a' (unused-argument)
mwe.py:1:6: W0612: Unused variable 'T' (unused-variable)
```

### Expected behavior

There should be no E0602, E0602, W0612.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
```

### OS / Environment

OS: Win32NT (10.0.19045.0)

**Comment by josmithua** [comment]
When running pylint on the above example I still get an error
```
pylint_test.py:2:30: E0602: Undefined variable 'T' (undefined-variable)
```
Using:
```
pylint 3.3.1
astroid 3.3.4
Python 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:26:36) [MSC v.1937 64 bit (AMD64)]
```

**Comment by inikishev** [comment]
I am still getting this, although for some reason sometimes it is fine and sometimes it shows UndefinedVariable, 
![Image](https://github.com/user-attachments/assets/494b4f37-a67a-404e-b68f-db4677936277)

**Comment by Pierre-Sassoulas** [comment]
Could be due to parallélisation that brings worse inférence if you analyse only part of the file (can't link to thé faq in the doc as I'm on mobile, sorry)
